scraper.py

```diff
@@ -127,64 +130,156 @@ def getLoans(self):
-            time.sleep(2)
-            if not chapter_title.get_attribute("textContent") in chapter_markers:
-                chapter_markers[chapter_title.get_attribute("textContent")] = timeline_current_time.get_attribute("textContent")
-            for i in range(2):
-                time_previous.click()
-                time.sleep(.15)
-            for i in range(4):
-                time_next.click()
-                time.sleep(.15)
-            time.sleep(.3)
-
-        urls = {}
 
+        # In order downloading of files
+        print("Getting files")
+
+        mp3_urls = self.requestsToMP3Files()
+        total_duration = 0
+        current_location = 0
+        part_num = 1
+        expected_duration = convert_metadata.to_seconds(expected_time)
+        missing_flag = False
+
+        # Main loop for walking through book
+        while True:
+            current_location = convert_metadata.to_seconds(timeline_current_time.get_attribute("textContent"))
+
+            # Skip to end of previously downloaded segment
+            while current_location < total_duration:
+
+                if not missing_flag:
+                    # Check if chapter is shorter than where the next part starts, and if so, can skip the whole chapter
+                    next_chapter_seconds = self.extract_minutes_to_seconds(chapter_next.get_attribute("textContent"))
+                    if total_duration-current_location > next_chapter_seconds:
+                        chapter_next.click()
+                        # print(f"Skipped chapter - {next_chapter_seconds} seconds")
+                        time.sleep(.5)
+                        current_location = convert_metadata.to_seconds(timeline_current_time.get_attribute("textContent"))
+
+                # # Jump by 15 seconds until we reach the end of the part
+                # time_next.click()
+
+                # Jump by 1 minute using page_down
+                self.driver.find_element(By.TAG_NAME, "body").send_keys(Keys.PAGE_DOWN)
+                current_location = convert_metadata.to_seconds(timeline_current_time.get_attribute("textContent"))
+
+            # Collect available urls, and download next part
+            mp3_urls = self.requestsToMP3Files()
+
+            url = mp3_urls.get(f"{part_num:02d}")
+            if not url:
+                print(f"Missing part {part_num}")
+                print("Trying to go back")
+                chapter_previous.click()
+                missing_flag = True
+                continue
+
+            length = overdrive_download.downloadMP3Part(url, f"{part_num:02d}", download_path, self.getCookies())
+
+            # If valid download, add the length of the part to the total, check progress through whole book
+            if length:
+                total_duration += length
+                print(f"{total_duration:.2f}/{expected_duration:.2f} sec  -  {total_duration/expected_duration*100:.2f}%")
+                if total_duration >= expected_duration*0.98:
```

Comment:
I would use the part numbers and the availability of next_chapter as the heuristic for completion here; there's very often a small credit chapter at the back of the book that might get broken out as a separate part sometimes, and this heuristic would usually skip it. I know I did it, but that was before I had part numbers :).